### ISSUE TITLE
Clang warns for unused Arm -march even when used by preprocessor

### ISSUE THREAD

**Issue by DavidSpickett** [llvm/llvm-project#55656]
(derived from https://github.com/ClangBuiltLinux/linux/issues/1315)

Summary: clang should not warn that compiler `-march` arguments are unused when the preprocessor is run on assembly files.

`/tmp/test.s` is just an empty file. Each time we're going to pass `-march` to the compiler and assembler with `-Wa`.
```
$ ./bin/clang --target=arm-linux-gnueabi -march=armv7-m -Wa,-march=armv7-m -Wunused-command-line-argument /tmp/test.s -c
clang-15: warning: argument unused during compilation: '-march=armv7-m' [-Wunused-command-line-argument]
```
If we assemble only then we are warned that the compiler `-march` is unused. So far so good.
```
$ ./bin/clang --target=arm-linux-gnueabi -march=armv7-m -Wa,-march=armv7-m -Wunused-command-line-argument -x assembler-with-cpp /tmp/test.s -c
clang-15: warning: argument unused during compilation: '-march=armv7-m' [-Wunused-command-line-argument]
```
If we assemble but apply the preprocessor first we are still warned that it is unused. Which isn't correct because this option actually sets some preprocessor defines (justification for that at the end).
```
$ ./bin/clang --target=arm-linux-gnueabi -march=armv7-m -Wa,-march=armv7-m -Wunused-command-line-argument /tmp/test.s -E
clang-15: warning: /tmp/test.s: 'assembler' input unused [-Wunused-command-line-argument]
clang-15: warning: argument unused during compilation: '-Wa,-march=armv7-m' [-Wunused-command-line-argument]
clang-15: warning: argument unused during compilation: '-Wunused-command-line-argument' [-Wunused-command-line-argument]
```
Now if we only preprocess an assembly file there's nothing to do so everything is unused. Fine.
```
$ ./bin/clang --target=arm-linux-gnueabi -march=armv7-m -Wa,-march=armv7-m -Wunused-command-line-argument -x assembler-with-cpp /tmp/test.s -E
# 1 "/tmp/test.s"
# 1 "<built-in>" 1
# 1 "/tmp/test.s" 2
```
If we run the preprocessor on the assembly file, everything is used. Part of me says that the `-Wa,-march` should be unused here. You are running the preprocessor on an assembly file but also we know that only the compiler `-march` is used to set defines. Not sure.

The key issue is that the compiler `-march` should be used if you're running the preprocessor. Clang should not be warning you that it is unused.

Since it governs defines like `__thumb2__` so clearly it does something. Here we see that removing the compiler `-march` causes this file to fail to assemble, therefore it is "used". If you put it back clang then tells you it is unused despite clearly being used to add the define.
```
$ cat /tmp/test2.s
#ifndef __thumb2__
#error
#endif
$ ./bin/clang --target=arm-linux-gnueabi -Wa,-march=armv7-m -Wunused-command-line-argument -x assembler-with-cpp /tmp/test2.s -c
/tmp/test2.s:2:2: error:
#error
 ^
1 error generated.
$ ./bin/clang --target=arm-linux-gnueabi -march=armv7-m -Wa,-march=armv7-m -Wunused-command-line-argument -x assembler-with-cpp /tmp/test2.
s -c
clang-15: warning: argument unused during compilation: '-march=armv7-m' [-Wunused-command-line-argument]
```
Somehow `-march` fixes the issue but is also unused.

**Comment by DavidSpickett** [comment]
The warning is emitted in `clang/lib/Driver/ToolChains/Arch/ARM.cpp::void arm::getARMTargetFeatures` and this probably applies to `-mcpu` as well.

```
  if (WaArch) {
    if (ArchArg)
      D.Diag(clang::diag::warn_drv_unused_argument)
          << ArchArg->getAsString(Args);
```

Unfortunately it's not as simple as checking if `ArchArg` has been "claimed" since it's claimed 3/4 times even if it's going to be overriden by `-Wa,-march`. This can happen when you search for an argument so it doesn't (at least in practice) mean it is used but that someone looked at it.

The spirit of the warning is "hey you gave us two architecture values but we're going to prefer one hopefully it's the right one". So "unused" makes sense as the type of warning.

We do get told if this is for the assembler but it's true either way. We can't tell if this is for the preprocessor and the assembler. I could add that context.

```
$ ./bin/clang --target=arm-linux-gnueabi -march=armv7-m -Wa,-march=armv7-m -Wunused-command-line-argument /tmp/test.s -c
Warning unused, claimed: 1 ForAS: 1
clang-15: warning: argument unused during compilation: '-march=armv7-m' [-Wunused-command-line-argument]
```
```
$ ./bin/clang --target=arm-linux-gnueabi -march=armv7-m -Wa,-march=armv7-m -Wunused-command-line-argument -x assembler-with-cpp /tmp/test.s -c
Warning unused, claimed: 1 ForAS: 1
clang-15: warning: argument unused during compilation: '-march=armv7-m' [-Wunused-command-line-argument]
```

**Comment by nickdesaulniers** [comment]
I think that
```
  if (WaArch) {
    if (ArchArg)
      D.Diag(clang::diag::warn_drv_unused_argument)
          << ArchArg->getAsString(Args);
```
just needs an additional check for `!getLangOpts().AsmPreprocessor`.

> The spirit of the warning is "hey you gave us two architecture values but we're going to prefer one hopefully it's the right one". So "unused" makes sense as the type of warning.

I guess you could do something odd like `-march=armv8 -Wa,-march=armv7`, which would be weird for .s, .S, or .c (with inline asm) files...

But I think for the kernel, the versions match. Maybe we could not warn if the versions match?

**Comment by DavidSpickett** [comment]
I'm not going to have time for this for the next few weeks. Anyone else feel free to pick it up.

**Comment by simonbcn** [comment]
Same problem on Arch when I try to compile `mpv` with `meson` using `clang`:  https://github.com/mpv-player/mpv/issues/11242

```
❯ clang --version
clang version 15.0.7
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
